File: src/orchestrator_agent/events/canvas_events.py
"""Event protocol emitted by the Capacity Orchestrator to the canvas.

This is the contract: every state change the canvas wants to render must be
expressible as one of these events. Adding a visual element requires adding
an event type.

Event IDs are ``str`` not ``uuid.UUID`` (per CLAUDE.md known gotcha — stdlib
``json.dumps`` can't serialize UUID even when Pydantic can). All event types
carry a ``type`` literal discriminator so the canvas reducer can dispatch
via tagged union.

Workflow nodes emit these by appending to ``ctx.state['canvas_events']``
via :func:`src.orchestrator_agent.events.emit.emit`. The A2A executor
(``runtime/agent_executor.py``) reads new entries on each ADK event and
forwards them as A2A ``Message`` events on the SSE stream.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Union

from pydantic import BaseModel, Field


def _new_id() -> str:
    """Generate a string UUID — never a ``uuid.UUID`` object.

    The deployed ADK runtime serializes structured output via stdlib
    ``json.dumps`` which doesn't know how to encode ``uuid.UUID``. See
    CLAUDE.md "UUID fields break ADK output serialization".
    """
    return str(uuid.uuid4())


class BaseEvent(BaseModel):
    """Common envelope for every canvas event."""

    event_id: str = Field(default_factory=_new_id)
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    workflow_id: str
    session_id: str


# --- Workflow lifecycle ----------------------------------------------------


class WorkflowStartedEvent(BaseEvent):
    type: Literal["workflow.started"] = "workflow.started"
    scenario: str  # "cargo-plane" | "buffer-planning"
    user_id: str
    initial_context: dict  # seed memory excerpt (per-persona)


class WorkflowCompletedEvent(BaseEvent):
    type: Literal["workflow.completed"] = "workflow.completed"
    final_output: dict  # SourcingPlan, BufferPlan, etc.
    duration_ms: int


class WorkflowFailedEvent(BaseEvent):
    type: Literal["workflow.failed"] = "workflow.failed"
    node_name: str
    error_class: str  # e.g. "TimeoutError"
    error_message: str


# --- Node lifecycle --------------------------------------------------------


class NodeStartedEvent(BaseEvent):
    type: Literal["node.started"] = "node.started"
    node_name: str
    node_kind: Literal["function", "llm", "router", "subworkflow"]


class NodeCompletedEvent(BaseEvent):
    type: Literal["node.completed"] = "node.completed"
    node_name: str
    duration_ms: int
    output_summary: str | None = None


# --- MCP & Knowledge Catalog calls (platform-visibility moments) ----------


class MCPCallStartedEvent(BaseEvent):
    type: Literal["mcp.call.started"] = "mcp.call.started"
    server: str  # "sap-mcp-server" | "maximo-mcp-server" | "fdp-mcp-server" | ...
    tool: str
    via_gateway: bool = True


class MCPCallCompletedEvent(BaseEvent):
    type: Literal["mcp.call.completed"] = "mcp.call.completed"
    server: str
    tool: str
    duration_ms: int
    result_summary: dict  # small payload — enough for canvas, not full response


class KnowledgeCatalogLookupEvent(BaseEvent):
    type: Literal["knowledge_catalog.lookup"] = "knowledge_catalog.lookup"
    canonical_id: str
    canonical_label: str
    aspects: dict  # full entity payload for the drawer


# --- Decisions -------------------------------------------------------------


class RouterDecisionEvent(BaseEvent):
    type: Literal["router.decision"] = "router.decision"
    router_name: str
    decision: str  # the routing key chosen
    rationale: str | None = None


# --- Cargo-plane specific events ------------------------------------------


class CapacityGapDetectedEvent(BaseEvent):
    type: Literal["capacity.gap_detected"] = "capacity.gap_detected"
    location: dict  # { latitude, longitude, label }
    canonical_asset_id: str
    deadline: str


class DoomedRouteProposedEvent(BaseEvent):
    """The naive 'cargo plane from Australia' plan, surfaced for contrast."""

    type: Literal["route.doomed_proposed"] = "route.doomed_proposed"
    from_location: dict
    to_location: dict
    estimated_cost_usd: float
    rationale: str


class EquivalentAssetFoundEvent(BaseEvent):
    """Equivalence node found a functional substitute."""

    type: Literal["equivalence.found"] = "equivalence.found"
    original_asset: dict
    equivalent_asset: dict
    confidence: float
    rationale_source: str
    location: dict


class RecommendedRouteFinalizedEvent(BaseEvent):
    type: Literal["route.recommended"] = "route.recommended"
    from_location: dict
    to_location: dict
    estimated_cost_usd: float
    avoided_cost_usd: float


# --- Buffer-planning specific events (declared now, used by sibling agents) -


class ForecastLoadedEvent(BaseEvent):
    type: Literal["forecast.loaded"] = "forecast.loaded"
    equipment_class: str
    timeline: list[dict]  # 12-week probabilistic forecast


class BufferRecommendationEvent(BaseEvent):
    type: Literal["buffer.recommendation"] = "buffer.recommendation"
    risk_tolerance: Literal["conservative", "balanced", "aggressive"]
    buffer_pct: float
    expected_idle_cost_usd: float
    expected_late_start_cost_usd: float
    on_time_probability: float


# --- Discriminated union for type-safe dispatch ---------------------------


CanvasEvent = Union[
    WorkflowStartedEvent,
    WorkflowCompletedEvent,
    WorkflowFailedEvent,
    NodeStartedEvent,
    NodeCompletedEvent,
    MCPCallStartedEvent,
    MCPCallCompletedEvent,
    KnowledgeCatalogLookupEvent,
    RouterDecisionEvent,
    CapacityGapDetectedEvent,
    DoomedRouteProposedEvent,
    EquivalentAssetFoundEvent,
    RecommendedRouteFinalizedEvent,
    ForecastLoadedEvent,
    BufferRecommendationEvent,
]

File: canvas/src/lib/canvas-events.ts
/**
 * canvas-events.ts
 *
 * TypeScript types mirroring the Pydantic schemas the Capacity Orchestrator
 * will emit on its A2A SSE stream (see
 * `src/orchestrator_agent/events/canvas_events.py`).
 *
 * Hand-authored against the TASK-10 spec — the matching Pydantic module is
 * being built in parallel by the backend agent. If the backend introduces
 * additional fields, update them here. The contract is enforced by the
 * deployed Orchestrator's `output_schema=`-style validation; the canvas
 * trusts the wire format and parses optimistically.
 *
 * Per the spec, all IDs are `string` (the backend uses `str` not
 * `uuid.UUID`; see CLAUDE.md known gotcha #6 — stdlib `json.dumps` can't
 * serialize `UUID` even when Pydantic can). Discriminated union via the
 * `type` literal.
 */

// ---------------------------------------------------------------------------
// Shared base envelope
// ---------------------------------------------------------------------------

export interface BaseEvent {
  event_id: string;
  /** ISO-8601 timestamp from the backend (datetime.utcnow().isoformat()). */
  timestamp: string;
  workflow_id: string;
  session_id: string;
}

// ---------------------------------------------------------------------------
// Geographic / asset payload shapes used across multiple event types
// ---------------------------------------------------------------------------

export interface EventLocation {
  latitude: number;
  longitude: number;
  label?: string;
}

export interface EventAsset {
  canonical_id: string;
  label: string;
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Workflow lifecycle
// ---------------------------------------------------------------------------

export interface WorkflowStartedEvent extends BaseEvent {
  type: "workflow.started";
  scenario: "cargo-plane" | "buffer-planning";
  user_id: string;
  initial_context: Record<string, unknown>;
}

export interface WorkflowCompletedEvent extends BaseEvent {
  type: "workflow.completed";
  final_output: Record<string, unknown>;
  duration_ms: number;
}

export interface WorkflowFailedEvent extends BaseEvent {
  type: "workflow.failed";
  node_name: string;
  error_class: string;
  error_message: string;
}

// ---------------------------------------------------------------------------
// Node lifecycle
// ---------------------------------------------------------------------------

export type NodeKind = "function" | "llm" | "router" | "subworkflow";

export interface NodeStartedEvent extends BaseEvent {
  type: "node.started";
  node_name: string;
  node_kind: NodeKind;
}

export interface NodeCompletedEvent extends BaseEvent {
  type: "node.completed";
  node_name: string;
  duration_ms: number;
  output_summary?: string | null;
}

// ---------------------------------------------------------------------------
// MCP & Knowledge Catalog calls (platform-visibility moments)
// ---------------------------------------------------------------------------

export interface MCPCallStartedEvent extends BaseEvent {
  type: "mcp.call.started";
  server: string;
  tool: string;
  via_gateway?: boolean;
}

export interface MCPCallCompletedEvent extends BaseEvent {
  type: "mcp.call.completed";
  server: string;
  tool: string;
  duration_ms: number;
  result_summary: Record<string, unknown>;
}

export interface KnowledgeCatalogLookupEvent extends BaseEvent {
  type: "knowledge_catalog.lookup";
  canonical_id: string;
  canonical_label: string;
  aspects: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Decisions
// ---------------------------------------------------------------------------

export interface RouterDecisionEvent extends BaseEvent {
  type: "router.decision";
  router_name: string;
  decision: string;
  rationale?: string | null;
}

// ---------------------------------------------------------------------------
// Cargo-plane specific events
// ---------------------------------------------------------------------------

export interface CapacityGapDetectedEvent extends BaseEvent {
  type: "capacity.gap_detected";
  location: EventLocation;
  canonical_asset_id: string;
  deadline: string;
}

export interface DoomedRouteProposedEvent extends BaseEvent {
  type: "route.doomed_proposed";
  from_location: EventLocation;
  to_location: EventLocation;
  estimated_cost_usd: number;
  rationale: string;
}

export interface EquivalentAssetFoundEvent extends BaseEvent {
  type: "equivalence.found";
  original_asset: EventAsset;
  equivalent_asset: EventAsset;
  confidence: number;
  rationale_source: string;
  location: EventLocation;
}

export interface RecommendedRouteFinalizedEvent extends BaseEvent {
  type: "route.recommended";
  from_location: EventLocation;
  to_location: EventLocation;
  estimated_cost_usd: number;
  avoided_cost_usd: number;
}

// ---------------------------------------------------------------------------
// Buffer-planning specific events
// ---------------------------------------------------------------------------

export interface ForecastLoadedEvent extends BaseEvent {
  type: "forecast.loaded";
  equipment_class: string;
  timeline: Array<Record<string, unknown>>;
}

export type BufferRiskTolerance = "conservative" | "balanced" | "aggressive";

export interface BufferRecommendationEvent extends BaseEvent {
  type: "buffer.recommendation";
  risk_tolerance: BufferRiskTolerance;
  buffer_pct: number;
  expected_idle_cost_usd: number;
  expected_late_start_cost_usd: number;
  on_time_probability: number;
}

// ---------------------------------------------------------------------------
// Discriminated union
// ---------------------------------------------------------------------------

export type CanvasEvent =
  | WorkflowStartedEvent
  | WorkflowCompletedEvent
  | WorkflowFailedEvent
  | NodeStartedEvent
  | NodeCompletedEvent
  | MCPCallStartedEvent
  | MCPCallCompletedEvent
  | KnowledgeCatalogLookupEvent
  | RouterDecisionEvent
  | CapacityGapDetectedEvent
  | DoomedRouteProposedEvent
  | EquivalentAssetFoundEvent
  | RecommendedRouteFinalizedEvent
  | ForecastLoadedEvent
  | BufferRecommendationEvent;

/** Convenient union of every `type` literal. */
export type CanvasEventType = CanvasEvent["type"];
